<a href="https://colab.research.google.com/github/valshie8231/05112005indira/blob/main/Salinan_dari_Salinan_dari_Pertemuan_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Manipulasi String

Proses mengubah-ubah struktur, format, atau isi dari sebuah string.
Berikut adalah beberapa contoh manipulasi string dalam python:


**Menggabungkan String**

In [ ]:
# Menggabungkan string
kata1 = "Halo"
kata2 = "Dunia"
kalimat_gabungan = kata1 + "  " + kata2
print("Kalimat gabungan:", kalimat_gabungan) # Output: Halo Dunia

Kalimat gabungan: Halo  Dunia


**Mengakses String**

In [ ]:
kalimat = "Belajar Python"
print("Karakter pertama:", kalimat[0])
print("Karakter kedua:", kalimat[1])
print("Karakter terakhir:", kalimat[-1])
print("Karakter kedua dari belakang:", kalimat[-2])

Karakter pertama: B
Karakter kedua: e
Karakter terakhir: n
Karakter kedua dari belakang: o


**Memeriksa Awal&Akhir Kalimat String**

In [ ]:
print("Memeriksa awalan string:", kalimat.startswith("Belajar")) # Output: True
print("Memeriksa akhiran string:", kalimat.endswith("Python")) # Output: True

Memeriksa awalan string: True
Memeriksa akhiran string: True


**Membalik String**

In [ ]:
# Membalik string
def balik_string(teks):
    return teks[::-1]

# Contoh penggunaan
print(balik_string("Halo Dunia"))


ainuD olaH


Penjelasan :
1. Sintaks [::-1] digunakan untuk mengambil seluruh karakter dalam string tetapi dalam urutan terbalik.


Mengubah Huruf Kapital

In [ ]:
ini_kata="halo dunia"
print(ini_kata)

halo dunia


In [ ]:
# Mengubah huruf kapital
def ubah_kapital(teks):
    return teks.upper(), teks.lower(), teks.title()

# Contoh penggunaan
upper, lower, title = ubah_kapital(ini_kata)
print("Uppercase:", upper)
print("Lowercase:", lower)
print("Titlecase:", title)


Uppercase: HALO DUNIA
Lowercase: halo dunia
Titlecase: Halo Dunia


Penjelasan:


1.   upper() mengubah semua huruf menjadi huruf besar.
2.   lower() mengubah semua huruf menjadi huruf kecil.
3.   title() mengubah huruf pertama setiap kata menjadi huruf besar.



**Menghapus Spasi di Awal/Akhir**

In [ ]:
print("   Halo Dunia   ")

   Halo Dunia   


In [ ]:
# Menghapus spasi di awal dan akhir
def hapus_spasi(teks):
    return teks.strip()

# Contoh penggunaan
print(hapus_spasi("   Halo Dunia   "))


Halo Dunia


Penjelasan:

1. Penjelasan: Metode strip() digunakan untuk menghapus spasi putih di awal dan akhir string.

**Mengganti Kata Di String**

In [ ]:
ganti_kata="Halo Dunia"
print("Kata Diawal Adalah",ganti_kata)

# Mengganti kata di string
def ganti_kata(teks, kata_lama, kata_baru):
    return teks.replace(kata_lama, kata_baru)

# Contoh penggunaan
print(ganti_kata("Halo Dunia", "Dunia", "Teman-Teman"))


Kata Diawal Adalah Halo Dunia
Halo Teman-Teman


Metode replace() mengganti semua kemunculan kata_lama dengan kata_baru dalam string.

**Memeriksa Substring Dalam String**

In [ ]:
# Memeriksa apakah substring ada dalam string
def cek_substring(teks, substring):
    return substring in teks

# Contoh penggunaan
print(cek_substring("Halo Dunia", "Dunia"))  # Output: True


True


Operator in memeriksa apakah substring ada dalam teks dan mengembalikan True atau False.

**Memecah String Menjadi List**

In [ ]:
# Memecah string menjadi list berdasarkan pemisah
def pecah_string(teks, pemisah=" "):
    return teks.split(pemisah)

# Contoh penggunaan
print(pecah_string("Halo Dunia Python"))


['Halo', 'Dunia', 'Python']


 Metode split() memisahkan string berdasarkan pemisah tertentu (default: spasi).

**Menggabungkan List menjadi String**

In [ ]:
Kata_kakgem=['Jikalau','Kamu','Sedih','Aku','Sedih']
print(Kata_kakgem)

['Jikalau', 'Kamu', 'Sedih', 'Aku', 'Sedih']


In [ ]:
def gabung_list(list_kata, pemisah=" "):
    return pemisah.join(list_kata)

# Contoh penggunaan
kalimat = gabung_list(Kata_kakgem)
print(kalimat)

kalimat_titik = gabung_list(Kata_kakgem, ".") #menggunakan titik sebagai pemisah
kalimat_titik

Jikalau Kamu Sedih Aku Sedih


'Jikalau.Kamu.Sedih.Aku.Sedih'

**Metode join() menggabungkan elemen dalam list menjadi string, dipisahkan oleh pemisah.**

**Mengubah String Menjadi Tuple**

In [ ]:
def string_ke_tuple(teks, pemisah=" "):
    return tuple(teks.split(pemisah))

# Contoh penggunaan
print(string_ke_tuple("Halo Dunia Python"))  # Output: ('Halo', 'Dunia', 'Python')
print(string_ke_tuple("1,2,3,4", ","))  # Output: ('1', '2', '3', '4')

('Halo', 'Dunia', 'Python')
('1', '2', '3', '4')


Penjelasan:
        teks: String yang akan diubah.
        pemisah: Pemisah untuk memecah string (default: spasi).

    Returns:
        Tuple yang berisi elemen-elemen dari string.
  

**Menghitung karakter atau kata**

In [ ]:
Nama_Terpanjang=" Hubert Blaine Wolfeschlegelsteinhausenbergerdorff"
print(Nama_Terpanjang)

 Hubert Blaine Wolfeschlegelsteinhausenbergerdorff


In [ ]:
def hitung_karakter(teks):
    return len(teks)

def hitung_kata(teks):
    kata = teks.split()
    return len(kata)

# Contoh penggunaan
jumlah_karakter = hitung_karakter(Nama_Terpanjang)
jumlah_kata = hitung_kata(Nama_Terpanjang)

print("Jumlah karakter:", jumlah_karakter)
print("Jumlah kata:", jumlah_kata)

Jumlah karakter: 50
Jumlah kata: 3


Penjelasan:

1. Fungsi hitung_karakter(teks)

*  Menggunakan len(teks) untuk menghitung jumlah semua karakter (termasuk spasi dan tanda baca) dalam string.

2. Fungsi hitung_kata(teks)

*   Menggunakan teks.split() untuk memecah string menjadi kata-kata berdasarkan spasi, lalu menghitung jumlah kata dengan len().







**Menghapus Karakter Khusus**

In [ ]:
ini_kalimatnya="$#Hai Saya Willy"
print(ini_kalimatnya)

$#Hai Saya Willy


In [ ]:
import re

def hapus_karakter_khusus(teks, karakter_khusus):
    teks_bersih = re.sub(f"[{re.escape(karakter_khusus)}]", "", teks)
    return teks_bersih

# Contoh penggunaan
karakter_khusus_yang_ingin_dihapus = "$#%"  # karakter yang ingin dihapus
kalimat_bersih = hapus_karakter_khusus(ini_kalimatnya, karakter_khusus_yang_ingin_dihapus)
print(kalimat_bersih)

Hai Saya Willy


Penjelasan:
Fungsi hapus_karakter_khusus(teks, karakter_khusus)


*   Tujuan: Menghapus karakter khusus tertentu dari string.
*   Cara Kerja:
1. re.escape(karakter_khusus): Melindungi karakter khusus agar diproses sebagai teks literal dalam regex.
2. re.sub(f"[{re.escape(karakter_khusus)}]", "", teks): Mengganti semua karakter yang cocok dengan kumpulan karakter_khusus dalam teks dengan string kosong (""), sehingga karakter tersebut dihapus.



**Membalik Urutan Kata**

In [ ]:
text_contoh="Halo Nama Saya Tegar"
print(text_contoh)

Halo Nama Saya Tegar


In [ ]:
def balik_kata(teks):
  kata = teks.split()
  balik = kata[::-1]
  return " ".join(balik)

  print(balik_kata(text_contoh))

# Slicing Pada String

In [ ]:
sliceing_string="Saya Sedang Belajar Python"
print(sliceing_string)

Saya Sedang Belajar Python


In [ ]:
# Contoh slicing
print(sliceing_string[:4])   # Mengambil karakter pertama hingga ke-4
print(sliceing_string[5:11]) # Mengambil karakter ke-5 hingga ke-10
print(sliceing_string[12:19]) # Mengambil karakter ke-12 hingga ke-18
print(sliceing_string[-6:])  # Mengambil 6 karakter terakhir
print(sliceing_string[::-1]) # Membalik seluruh string

NameError: name 'sliceing_string' is not defined

# Concatenation String

In [ ]:
# Variabel string
kata1 = "Belajar"
kata2 = "Python"
kata3 = "Sangat Menyenangkan"

# Concatenation menggunakan operator +
hasil = kata1 + " " + kata2 + " " + kata3
print(hasil)

# Concatenation menggunakan f-string (format string)
hasil_fstring = f"{kata1} {kata2} {kata3}"
print(hasil_fstring)

# Concatenation menggunakan method join()
daftar_kata = [kata1, kata2, kata3]
hasil_join = " ".join(daftar_kata)
print(hasil_join)


Belajar Python Sangat Menyenangkan
Belajar Python Sangat Menyenangkan
Belajar Python Sangat Menyenangkan


# Library  Python

**NUMPY**
NumPy (Numerical Python) adalah library Python untuk komputasi ilmiah. Digunakan untuk:


*   Operasi matematis pada array multidimensi.
*   Manipulasi data numerik.
*   Pengolahan data dengan performa tinggi.






In [ ]:
import numpy as np

# Fungsi trigonometri
angle = np.pi / 4  # 45 derajat dalam radian
sin_value = np.sin(angle)
cos_value = np.cos(angle)

print(f"Sin(45°): {sin_value}")
print(f"Cos(45°): {cos_value}")

# Fungsi eksponensial dan logaritma
exp_value = np.exp(2)  # e^2
log_value = np.log(10) # log natural

print(f"e^2: {exp_value}")
print(f"Log(10): {log_value}")


Sin(45°): 0.7071067811865475
Cos(45°): 0.7071067811865476
e^2: 7.38905609893065
Log(10): 2.302585092994046


Penjelasan:
1. Fungsi Trigonometri (np.sin() dan np.cos()):


*   np.pi / 4: Menghitung sudut 45 derajat dalam bentuk radian (karena NumPy menggunakan radian untuk perhitungan trigonometri).
*   np.sin(angle): Menghitung nilai sinus dari sudut 45 derajat.
*   np.cos(angle): Menghitung nilai cosinus dari sudut 45 derajat.

2. Fungsi Eksponensial dan Logaritma:


*   np.exp(2): Menghitung eksponensial dari angka 2, yaitu e^2 (nilai sekitar 7.389).
*   np.log(10): Menghitung logaritma natural dari 10 (basis e), yang menghasilkan nilai sekitar 2.3026.







**PANDAS**
Pandas adalah pustaka Python untuk analisis dan manipulasi data. Mendukung operasi data berbasis tabel dan time-series.
Digunakan untuk:


*   Membaca dan menulis data dari berbagai format (CSV, Excel, SQL, JSON, dll.).
*   Membersihkan, mengolah, dan menganalisis data dengan efisien.



In [ ]:
import pandas as pd

# Membuat DataFrame sederhana
data = {'Nama': ['Alice', 'Bob', 'Charlie'],
        'Usia': [25, 30, 28],
        'Kota': ['New York', 'London', 'Paris']}
df = pd.DataFrame(data)

# Menampilkan DataFrame
print(df)

# Mengakses kolom tertentu
print(df['Nama'])

# Mengakses baris tertentu
print(df.loc[0]) #Menggunakan label baris

print(df.iloc[0]) #Menggunakan index baris

      Nama  Usia      Kota
0    Alice    25  New York
1      Bob    30    London
2  Charlie    28     Paris
0      Alice
1        Bob
2    Charlie
Name: Nama, dtype: object
Nama       Alice
Usia          25
Kota    New York
Name: 0, dtype: object
Nama       Alice
Usia          25
Kota    New York
Name: 0, dtype: object


Penjelasan:


*   pd.DataFrame(): Digunakan untuk membuat DataFrame dari data yang disediakan dalam bentuk dictionary atau struktur lainnya.
*   df['Nama']: Mengakses kolom tertentu (dalam hal ini kolom 'Nama').
*   df.loc[]: Mengakses baris menggunakan label baris.
*   df.iloc[]: Mengakses baris menggunakan index numerik.





# Praktek Time

**Soal No 1**

Manipulasi String dengan Slicing

kalimat = "Hari Ini Adalah Hari Ketiga Saya Mengikuti Kegiatan Ngobar"



*   Ambil dan tampilkan 6 karakter pertama dari string tersebut.
*   Ambil dan tampilkan 5 karakter terakhir dari string tersebut.

*   Ambil dan tampilkan kata "Ketiga" dari string tersebut.
*   Tampilkan string tersebut dengan urutan terbalik.






**Soal No 2**

Concatenation String

Buatlah 3 Variabel String (Kalimatnya Terserah) Lalu Kerjakan:


*   Gabungkan ketiga Variabel menggunakan operator + dan tampilkan hasilnya.
*   Gabungkan ketiga Variabel menggunakan f-string dan tampilkan hasilnya.



*   Gabungkan ketiga kata di atas menggunakan method .join() dan tampilkan hasilnya.


**Soal No 3**
Menghitung Jumlah Karakter dan Kata

tex_soal="aku adalah seorang kapten dari sebuah tim sepak bola liga 3 liga belgia"

yang harus dilakukan:


*   Hitung dan tampilkan jumlah karakter dalam string tersebut.
*   Hitung dan tampilkan jumlah kata dalam string tersebut.



**Soal No 4**

Hitunglah nilai rata-rata, nilai minimum, dan nilai maksimum dari array NumPy berikut:

import numpy as np

arr = np.array([1, 5, 2, 9, 7, 3])

Yang Harus Dilakukan

1. Import library NumPy dan buat array seperti yang ditunjukkan di atas.
2. Gunakan fungsi-fungsi NumPy yang sesuai untuk menghitung nilai rata-rata (mean), nilai minimum (min), dan nilai maksimum (max) dari array tersebut.
3. Tampilkan hasil perhitungan tersebut.
Petunjuk: